hudi-common/src/main/java/org/apache/hudi/avro/GenericAvroSerializer.java

```diff
@@ -0,0 +1,148 @@
+ * <p>
+ * Heavily adapted from:
+ * <p>
+ * <a href="https://github.com/apache/spark/blob/master/core/src/main/scala/org/apache/spark/serializer/GenericAvroSerializer.scala">GenericAvroSerializer.scala</a>
+ * <p>
+ * As {@link org.apache.hudi.common.util.SerializationUtils} is not shared between threads and does not concern any
+ * shuffling operations, compression and decompression cache is omitted as network IO is not a concern.
+ * <p>
+ * Unlike Spark's implementation, the class and constructor is not initialized with a predefined map of avro schemas.
+ * This is the case as schemas to read and write are not known beforehand.
+ *
+ * @param <D> the subtype of [[GenericContainer]] handled by this serializer
+ */
+public class GenericAvroSerializer<D extends GenericContainer> extends Serializer<D> {
+
+  // reuses the same datum reader/writer since the same schema will be used many times
+  private final HashMap<Schema, DatumWriter<D>> writerCache = new HashMap<>();
+  private final HashMap<Schema, DatumReader<D>> readerCache = new HashMap<>();
+
+  // fingerprinting is very expensive so this alleviates most of the work
+  private final HashMap<Schema, Long> fingerprintCache = new HashMap<>();
+  private final HashMap<Long, Schema> schemaCache = new HashMap<>();
+
+  private Long getFingerprint(Schema schema) {
+    if (fingerprintCache.containsKey(schema)) {
+      return fingerprintCache.get(schema);
+    } else {
+      Long fingerprint = SchemaNormalization.parsingFingerprint64(schema);
+      fingerprintCache.put(schema, fingerprint);
+      return fingerprint;
+    }
+  }
+
+  private Schema getSchema(Long fingerprint, byte[] schemaBytes) {
+    if (schemaCache.containsKey(fingerprint)) {
+      return schemaCache.get(fingerprint);
+    } else {
+      String schema = new String(schemaBytes, StandardCharsets.UTF_8);
+      Schema parsedSchema = new Schema.Parser().parse(schema);
+      schemaCache.put(fingerprint, parsedSchema);
+      return parsedSchema;
+    }
+  }
+
+  private DatumWriter<D> getDatumWriter(Schema schema) {
+    DatumWriter<D> writer;
+    if (writerCache.containsKey(schema)) {
+      writer = writerCache.get(schema);
+    } else {
+      writer = new GenericDatumWriter<>(schema);
+      writerCache.put(schema, writer);
+    }
+    return writer;
+  }
+
+  private DatumReader<D> getDatumReader(Schema schema) {
+    DatumReader<D> reader;
+    if (readerCache.containsKey(schema)) {
+      reader = readerCache.get(schema);
+    } else {
+      reader = new GenericDatumReader<>(schema);
+      readerCache.put(schema, reader);
+    }
+    return reader;
+  }
+
+  private void serializeDatum(D datum, Output output) throws IOException {
+    Encoder encoder = EncoderFactory.get().directBinaryEncoder(output, null);
+    Schema schema = datum.getSchema();
+    Long fingerprint = this.getFingerprint(schema);
+    byte[] schemaBytes = schema.toString().getBytes(StandardCharsets.UTF_8);
```

Comment:
This is Spark impl for serializing:

```scala
  def serializeDatum(datum: D, output: KryoOutput): Unit = {
    val encoder = EncoderFactory.get.binaryEncoder(output, null)
    val schema = datum.getSchema
    val fingerprint = fingerprintCache.getOrElseUpdate(schema, {
      SchemaNormalization.parsingFingerprint64(schema)
    })
    schemas.get(fingerprint) match {
      case Some(_) =>
        output.writeBoolean(true)
        output.writeLong(fingerprint)
      case None =>
        output.writeBoolean(false)
        val compressedSchema = compress(schema)
        output.writeInt(compressedSchema.length)
        output.writeBytes(compressedSchema)
    }

    writerCache.getOrElseUpdate(schema, GenericData.get.createDatumWriter(schema))
      .asInstanceOf[DatumWriter[D]]
      .write(datum, encoder)
    encoder.flush()
  }
```

So what's the point of the fingerprint here if we always serialize the schema bytes?

Reply:
Although we are ALWAYS serializing the bytes, it may or may not be parsed by:

```java
new Schema.Parser().parse(schema)
```

The fingerprint here is used as a key to check if parsing of the bytes -> Schema is required, if it is not required, it will skip this step.

Caching and updating of cache is done in the methods:
1. getFingerprint
2. getSchema

Reply:
If you refer to the Spark's implementation, `schemaCache` is always a subset of `schemas` (which is passed avro schemas registered by the session when constructing the GenericAvroSerializer).

As such, if Spark's `GenericAvroSerializer` is given a schema that is not in `schemas`, it will always serialize/write the compressed bytes.

Reply:
Then at least there is no need to serialize the fingerprint right? Just deserialize from the schema bytes directly should be fine.

Reply:
> Then at least there is no need to serialize the fingerprint right? Just deserialize from the schema bytes directly should be fine.

So, `schemaCache` will be keyed by byte[]?

```java
private final HashMap<byte[], Schema> schemaCache = new HashMap<>();
```

Reply:
The schema cache could be useless now, see the code in Spark:

```scala
    def deserializeDatum(input: KryoInput): D = {
    val schema = {
      if (input.readBoolean()) {
        val fingerprint = input.readLong()
        schemaCache.getOrElseUpdate(fingerprint, {
          schemas.get(fingerprint) match {
            case Some(s) => new Schema.Parser().setValidateDefaults(false).parse(s)
            case None =>
              throw new SparkException(
                "Error reading attempting to read avro data -- encountered an unknown " +
                  s"fingerprint: $fingerprint, not sure what schema to use.  This could happen " +
                  "if you registered additional schemas after starting your spark context.")
          }
        })
      } else {
        val length = input.readInt()
        decompress(ByteBuffer.wrap(input.readBytes(length)))
      }
    }
    val decoder = DecoderFactory.get.directBinaryDecoder(input, null)
    readerCache.getOrElseUpdate(schema, GenericData.get.createDatumReader(schema))
      .asInstanceOf[DatumReader[D]]
      .read(null.asInstanceOf[D], decoder)
  }
```

When fingleprint is disabled, the schema is decoded directly from the compressee bytes.

Reply:
However, compressBytes and decompressBytes are cached right?

Reply:
> When fingleprint is disabled, the schema is decoded directly from the compressee bytes.

Yeap. In this PR, our `schemaCache` serves the purpose of the `decompressCache`, which is the cache used in the `decompress` function.

If you look at Spark's implementation, the schema is not ALWAYS decoded directly from compressed bytes.

If the compressed bytes have been decoded before, it will be skipped as such:

```scala
decompressCache.getOrElseUpdate... 
```

Please refer to the full snippet that I have pasted below.

# Our PR
```java
 private Schema getSchema(Long fingerprint, byte[] schemaBytes) {
    if (schemaCache.containsKey(fingerprint)) {
      return schemaCache.get(fingerprint);
    } else {
      String schema = new String(schemaBytes, StandardCharsets.UTF_8);
      Schema parsedSchema = new Schema.Parser().parse(schema);
      schemaCache.put(fingerprint, parsedSchema);
      return parsedSchema;
    }
  }

  private D deserializeDatum(Input input) throws IOException {
    Long fingerprint = input.readLong();
    int schemaBytesLen = input.readInt();
    byte[] schemaBytes = input.readBytes(schemaBytesLen);
    Schema schema = getSchema(fingerprint, schemaBytes);
    Decoder decoder = DecoderFactory.get().directBinaryDecoder(input, null);
    return getDatumReader(schema).read(null, decoder);
  }
```


# Spark's implementation

```scala
  def decompress(schemaBytes: ByteBuffer): Schema = decompressCache.getOrElseUpdate(schemaBytes, {
    val bis = new ByteArrayInputStream(
      schemaBytes.array(),
      schemaBytes.arrayOffset() + schemaBytes.position(),
      schemaBytes.remaining())
    val in = codec.compressedInputStream(bis)
    val bytes = Utils.tryWithSafeFinally {
      IOUtils.toByteArray(in)
    } {
      in.close()
    }
    new Schema.Parser().parse(new String(bytes, StandardCharsets.UTF_8))
  })
```

Reply:
You are right, we can use the schema bytes as the cache key. But be caution of the `ByteBuffer` equals impls.

Reply:
In that case, I'll change the cache as such?

```
private final HashMap<Schema, byte[]> fingerprintCache = new HashMap<>();
private final HashMap<byte[], Schema> schemaCache = new HashMap<>();
```

Reply:
Done

Reply:
Also, something to consider, IIUC, serializing the the (Long) fingerprints allows for faster deserialization reading. 

```java
@Test
public void testRocksDbDiskMapPutDecimal() throws IOException {
  RocksDbDiskMap<String, HoodieRecord> rocksDbBasedMap = new RocksDbDiskMap<>(basePath);
  HoodieRecord avroRecord = createAvroRecordWithDecimalOrderingField();
  rocksDbBasedMap.put(avroRecord.getRecordKey(), avroRecord);
  assertDoesNotThrow(() -> rocksDbBasedMap.get(avroRecord.getRecordKey()));
  
  long start;
  for (int h = 0; h < 5; h++) {
    start = System.currentTimeMillis();
    for (int i = 0; i < 5999999; i++) {
      rocksDbBasedMap.get(avroRecord.getRecordKey());
    }
    System.out.println(System.currentTimeMillis() - start);
  }
}
```

```
Using HashMap<ByteBuffer, Schema> cache
27746
23477
20766
20541
20247

Using HashMap<Long, Schema> cache
26384
20606
17815
16544
16070
```

Reply:
Yeah, a classic CPU and storage trade off.

Reply:
Yeah, given that we are writing a long as a fingerprint, the storage penalty seems like a reasonable tradeoff for speed...